File: loglib-java/src/main/java/com/robog/loglib/LogBeanPool.java
package com.robog.loglib;

/**
 * Created by yuxingdong on 2018/6/23.
 */
public final class LogBeanPool {

    private static final Object sPoolSync = new Object();
    private static LogBean sPool;
    private static int sPoolSize = 0;
    private static final int MAX_POOL_SIZE = 60;

    public static LogBean get() {
        synchronized (sPoolSync) {
            if (sPool != null) {
                LogBean bean = sPool;
                sPool = bean.next;
                bean.next = null;
                sPoolSize--;
                return bean;
            }
        }
        return new LogBean();
    }

    static void recycle(LogBean logBean) {

        logBean.priority = 0;
        logBean.tag = "";
        logBean.message = "";
        logBean.time = "";
        logBean.stackTrace = "";

        synchronized (sPoolSync) {
            if (sPoolSize < MAX_POOL_SIZE) {
                logBean.next = sPool;
                sPool = logBean;
                sPoolSize++;
            }
        }
    }
}

File: loglib-kotlin/src/main/java/com/sunyard/loglib/LogBeanPool.kt
package com.sunyard.loglib

/**
 * Created by yuxingdong on 2018/6/29.
 */
object LogBeanPool {

    private val poolSync = Any()
    private var pool: LogBean? = null
    private var poolSize = 0
    private val MAX_POOL_SIZE = 60

    fun get(): LogBean {
        synchronized(poolSync) {
            if (pool != null) {
                val bean = pool
                pool = bean!!.next
                bean.next = null
                poolSize--
                return bean
            }
        }
        return LogBean(next = null)
    }

    internal fun recycle(logBean: LogBean) {

        logBean.priority = 0
        logBean.tag = ""
        logBean.message = ""
        logBean.time = ""
        logBean.stackTrace = ""

        synchronized(poolSync) {
            if (poolSize < MAX_POOL_SIZE) {
                logBean.next = pool
                pool = logBean
                poolSize++
            }
        }
    }
}
